File: content/snippets/python/interview_lcci/02/07_intersection_of_two_linked_lists.py
class ListNode:
    def __init__(self, val: int, next: "ListNode | None" = None) -> None:
        self.val = val
        self.next = next


def build_list(nums: list[int]) -> ListNode | None:
    dummy = ListNode(0)
    tail = dummy
    for n in nums:
        tail.next = ListNode(n)
        tail = tail.next
    return dummy.next


def get_intersection_node(a: ListNode | None, b: ListNode | None) -> ListNode | None:
    if a is None or b is None:
        return None

    p1, p2 = a, b
    while p1 is not p2:
        p1 = p1.next if p1 is not None else b
        p2 = p2.next if p2 is not None else a
    return p1


def main() -> None:
    common = build_list([8, 10])

    a = build_list([3, 1, 5, 9])
    b = build_list([4, 6])

    tail_a = a
    while tail_a is not None and tail_a.next is not None:
        tail_a = tail_a.next
    if tail_a is not None:
        tail_a.next = common

    tail_b = b
    while tail_b is not None and tail_b.next is not None:
        tail_b = tail_b.next
    if tail_b is not None:
        tail_b.next = common

    node = get_intersection_node(a, b)
    print(node.val if node else "null")


if __name__ == "__main__":
    main()

File: content/snippets/typescript/interview_lcci/02/07_intersection_of_two_linked_lists.ts
class ListNode {
  val: number;
  next: ListNode | null;

  constructor(val: number, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function buildList(nums: number[]): ListNode | null {
  const dummy = new ListNode(0);
  let tail = dummy;
  for (const n of nums) {
    tail.next = new ListNode(n);
    tail = tail.next;
  }
  return dummy.next;
}

function getIntersectionNode(a: ListNode | null, b: ListNode | null): ListNode | null {
  if (a === null || b === null) {
    return null;
  }

  let p1: ListNode | null = a;
  let p2: ListNode | null = b;
  while (p1 !== p2) {
    p1 = p1 !== null ? p1.next : b;
    p2 = p2 !== null ? p2.next : a;
  }
  return p1;
}

function main(): void {
  const common = buildList([8, 10]);
  const a = buildList([3, 1, 5, 9]);
  const b = buildList([4, 6]);

  let tailA = a;
  while (tailA !== null && tailA.next !== null) {
    tailA = tailA.next;
  }
  if (tailA !== null) {
    tailA.next = common;
  }

  let tailB = b;
  while (tailB !== null && tailB.next !== null) {
    tailB = tailB.next;
  }
  if (tailB !== null) {
    tailB.next = common;
  }

  const node = getIntersectionNode(a, b);
  console.log(node ? node.val : "null");
}

main();
